service/src/DCGMDevicePool.hpp

```diff
@@ -0,0 +1,99 @@
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY LOG OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef DCGMDEVICEPOOL_HPP_INCLUDE
+#define DCGMDEVICEPOOL_HPP_INCLUDE
+
+namespace geopm
+{
+    /// An interface for the NVIDIA Data Center GPU Manager (DCGM).
+    /// This class is a wrapper around all calls to the DCGM library
+    /// and is intented to be called via the DCGMIOGroup.  Its primary
+    /// function is provided an abstracted interface to DCGM metrics
+    /// of interest.
+    class DCGMDevicePool
+    {
+        public:
+            enum geopm_dcgm_field_ids_e {
+                /*!
+                 * @brief Field ID associated with DCGM SM Active metrics
+                 */
+                M_SM_ACTIVE,
+                /*!
+                 * @brief Field ID associated with SM Occupancy metrics
+                 */
+                M_SM_OCCUPANCY,
+                /*!
+                 * @brief Field ID associated with DCGM DRAM Active metrics
+                 */
+                M_DRAM_ACTIVE,
+                /*!
+                 * @brief Number of valid field ids
+                 */
+                M_FIELD_IDS
+            };
+
+            DCGMDevicePool() = default;
+            virtual ~DCGMDevicePool() = default;
+
+            /// @brief Number of accelerators that support DCGM on the platform.
+            /// @return Number of accelerators.
+            virtual int dcgm_device() const = 0;
+            /// @brief Get the value for the provided geopm_field_id.
+            ///        This value should not change unless update_field_value
+            //         has been called
+            /// @param [in] accel_idx The index indicating a particular
+            ///        accelerator.
+            /// @return The value for the specified field
+            virtual double sample_field_value(int accel_idx, int geopm_field_id) const = 0;
+
+            /// @brief Query DCGM for the latest value for an accelerator.
+            ///        Note that this is the last value DCGM cahced.  This
+            //         updates the DCGM device pool stored value that is provided
+            //         via the sample_field_value function
+            /// @param [in] accel_idx The index indicating a particular
+            ///        accelerator.
+            virtual void update_field_value(int accel_idx) = 0;
+            /// @brief Set field update rate for DCGM devices.
+            /// @param [in] field_update_rate DCGM update rate in microseconds.
+            virtual void field_update_rate(int field_update_rate) = 0;
+            /// @brief Set maximum storage time for for DCGM devices.
+            /// @param [in] max_storage_time maximum storage time in seconds
+            virtual void max_storage_time(int max_storage_time) = 0;
```

Comment:
I don't understand what "storage_time" means

Reply:
Storage time is defined by the [DCGM API](https://docs.nvidia.com/datacenter/dcgm/1.6/dcgm-api/group__DCGMAPI__FI.html#group__DCGMAPI__FI_1gd04fa06405af6165fd85310dee9eae2a), it's how long the DCGM process will store those samples before dropping them.